hivemind/utils/crypto.py

```diff
@@ -0,0 +1,84 @@
+import re
+from dataclasses import astuple, dataclass
+from typing import Optional
+
+from cryptography.exceptions import InvalidSignature
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import padding, rsa
+
+from hivemind.utils.serializer import MSGPackSerializer
+
+
+@dataclass(init=True, repr=True, frozen=True)
+class ProtectedRecord:
+    key: bytes
+    subkey: bytes
+    value: bytes
+    expiration_time: float
+
+
+class RecordValidatorBase:
+    def validate(self, record: ProtectedRecord) -> bool:
+        raise NotImplementedError()
+
+    def sign(self, record: ProtectedRecord) -> bytes:
+        raise NotImplementedError()
+
+
+class RSASignatureValidator(RecordValidatorBase):
+    """
+    Introduces a notion of *protected records* whose key/subkey contains substring
+    "[owner:ssh-rsa ...]" (the format can be changed) with an RSA public key of the owner.
+
+    If this validator is used, changes to such records always must be signed with
+    the corresponding private key (so only the owner can change them).
+    """
+
+    def __init__(self, marker_format: bytes=b'[owner:key]'):
+        self._marker_format = marker_format
+        self._marker_re = re.compile(re.escape(marker_format).replace(b'key', rb'(.+?)'))
+
+        self._private_key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
+
+        serialized_public_key = self._private_key.public_key().public_bytes(
+            encoding=serialization.Encoding.OpenSSH, format=serialization.PublicFormat.OpenSSH)
+        self._our_marker = marker_format.replace(b'key', serialized_public_key)
+
+        self._padding = padding.PSS(mgf=padding.MGF1(hashes.SHA256()),
+                                    salt_length=padding.PSS.MAX_LENGTH)
+        self._hash_algorithm = hashes.SHA256()
+
+    @property
+    def ownership_marker(self) -> bytes:
+        return self._our_marker
+
+    def validate(self, record: ProtectedRecord, signature: Optional[bytes]) -> None:
+        self._marker_re.findall(record.key)
+
+        public_keys = self._marker_re.findall(record.key)
+        if record.subkey is not None:
+            public_keys += self._marker_re.findall(record.subkey)
+        if not public_keys:
+            return
+
+        if signature is None:
+            raise InvalidSignature('Signature is required but not present')
+        if len(set(public_keys)) > 1:
+            raise InvalidSignature("Key and subkey can't contain different public keys")
+        try:
+            public_key = serialization.load_ssh_public_key(public_keys[0])
+        except ValueError as e:
+            raise InvalidSignature(f'Failed to deserialize public key: {e}')
+
+        public_key.verify(signature, self._serialize_record(record),
```

Comment:
Don't we need to return result of verify()?

Reply:
[RSAPublicKey.verify](https://cryptography.io/en/latest/hazmat/primitives/asymmetric/rsa.html#cryptography.hazmat.primitives.asymmetric.rsa.RSAPublicKey.verify) returns None if the signature is valid and throws `InvalidSignature` if it is not.

I will fix the code to make this behavior more clear for a reader.

Reply:
Done